wfe2/wfe_test.go

```diff
@@ -3327,34 +3384,117 @@ func TestNewAccountWhenGetRegByKeyNotFound(t *testing.T) {
+		test.AssertEquals(t, chall.Status, core.StatusInvalid)
+	}
+}
+
+func TestPrepWildcardAuthzForDisplay(t *testing.T) {
+	t.Parallel()
 	wfe, _, _ := setupWFE(t)
 
-	// Make an authz for a wildcard identifier
 	authz := &core.Authorization{
 		ID:             "12345",
 		Status:         core.StatusPending,
 		RegistrationID: 1,
 		Identifier:     identifier.DNSIdentifier("*.example.com"),
 		Challenges: []core.Challenge{
-			{
-				Type: "dns",
-			},
+			{Type: core.ChallengeTypeDNS01, Status: core.StatusPending, Token: "token"},
 		},
 	}
 
-	// Prep the wildcard authz for display
+	// This modifies the authz in-place.
 	wfe.prepAuthorizationForDisplay(&http.Request{Host: "localhost"}, authz)
 
-	// The authz should not have a wildcard prefix in the identifier value
+	// The identifier should not start with a star, but the authz should be marked
+	// as a wildcard.
 	test.AssertEquals(t, strings.HasPrefix(authz.Identifier.Value, "*."), false)
-	// The authz should be marked as corresponding to a wildcard name
 	test.AssertEquals(t, authz.Wildcard, true)
+}
 
-	// We expect the authz challenge has its URL set and the URI emptied.
-	authz.ID = "12345"
-	wfe.prepAuthorizationForDisplay(&http.Request{Host: "localhost"}, authz)
-	chal := authz.Challenges[0]
-	test.AssertEquals(t, chal.URL, "http://localhost/acme/chall-v3/12345/po1V2w")
+func TestPrepAuthzForDisplayShuffle(t *testing.T) {
+	t.Parallel()
+	wfe, _, _ := setupWFE(t)
+
+	authz := &core.Authorization{
+		ID:             "12345",
+		Status:         core.StatusPending,
+		RegistrationID: 1,
+		Identifier:     identifier.DNSIdentifier("example.com"),
+		Challenges: []core.Challenge{
+			{Type: core.ChallengeTypeDNS01, Status: core.StatusPending, Token: "token"},
+			{Type: core.ChallengeTypeHTTP01, Status: core.StatusPending, Token: "token"},
+			{Type: core.ChallengeTypeTLSALPN01, Status: core.StatusPending, Token: "token"},
+		},
+	}
+
+	// The challenges should be presented in an unpredictable order.
+
+	// Create a structure to count how many times each challenge type ends up in
+	// each position in the output authz.Challenges list.
+	counts := make(map[core.AcmeChallenge]map[int]int)
+	counts[core.ChallengeTypeDNS01] = map[int]int{0: 0, 1: 0, 2: 0}
+	counts[core.ChallengeTypeHTTP01] = map[int]int{0: 0, 1: 0, 2: 0}
+	counts[core.ChallengeTypeTLSALPN01] = map[int]int{0: 0, 1: 0, 2: 0}
+
+	// Prep the authz 100 times, and count where each challenge ended up each time.
+	for range 100 {
+		// This modifies the authz in place
+		wfe.prepAuthorizationForDisplay(&http.Request{Host: "localhost"}, authz)
+		for i, chall := range authz.Challenges {
+			counts[chall.Type][i] += 1
+		}
+	}
+
+	// Ensure that at least some amount of randomization is happening.
+	for challType, indices := range counts {
+		for index, count := range indices {
+			test.Assert(t, count > 10, fmt.Sprintf("challenge type %s did not appear in position %d as often as expected", challType, index))
+		}
+	}
```

Comment:
I just wanted to toy with this test a bit and here's where I ended up. I increased the runs to 1000 to increase statistical confidence (at very little cost). Refactored the check to be more specific, detailing exactly where any gaps occur. Feel free to take it or leave it though.

```suggestion
	// Prep the authz 1000 times, and count where each challenge ended up each time.
	trials := 1000
	for range trials {
		// This modifies the authz in place
		wfe.prepAuthorizationForDisplay(&http.Request{Host: "localhost"}, authz)
		for i, chall := range authz.Challenges {
			counts[chall.Type][i] += 1
		}
	}

	// Check for roughly uniform distribution across positions, ensuring each is
	// hit at least half the expected number of times.
	expectedCount := trials / len(authz.Challenges)
	for challType, positionCounts := range counts {
		for position, count := range positionCounts {
			if count < expectedCount/2 {
				t.Errorf("challenge type %s was at position %d, appearing %d times, expected at least %d times",
					challType, position, count, expectedCount/2)
			}
		}
	}
```

Reply:
Per discussion offline, we're confident that this test at 100 rounds is sufficiently strong against flaky failure, and we've updated with WFE to use the built-in `rand.Shuffle` so we don't need to have our own confidence in the distribution.